akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -359,8 +364,8 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
             outstandingConnectionLevelWindow += windowSizeIncrement
           }
 
-          buffer.onDataFrame(d).getOrElse(
-            maybeFinishStream(d.endStream))
+          buffer.onDataFrame(d)
```

Comment:
The main change is that, before, we moved out of the state receiving data when we received the frame with `END_STREAM` which is right if we watch the states from the outside. However, internally, the final frames might not yet have been dispatched in which case we would have dropped the buffer on the floor and leaked it in cases where the connection was aborted in this state.

Alternatively we could have spelled this out as another state (END_STREAM received but not yet all data dispatched), however, since we already have a state matrix with the product of the state of all the four ends of each stream (network and user in and out), this would have easily let to a multitude of new states. That's why I went with the slightly impure solution of extending the meaning of receiving states until after the user handler has actually pulled all the data.